File: problems/07_trees/0100_same_tree/solution.swift
// Problem: 0100. Same Tree
// Difficulty: Easy

import Foundation

class TreeNode {
    var val: Int; var left: TreeNode?; var right: TreeNode?
    init(_ v: Int = 0, _ l: TreeNode? = nil, _ r: TreeNode? = nil) { val = v; left = l; right = r }
}

func fromLevel(_ a: [Int?]) -> TreeNode? {
    if a.isEmpty { return nil }
    let root = TreeNode(a[0]!); var q = [root]; var qi = 0, i = 1
    while qi < q.count && i < a.count {
        let n = q[qi]; qi += 1
        if i < a.count, let v = a[i] { n.left = TreeNode(v); q.append(n.left!) }
        i += 1
        if i < a.count, let v = a[i] { n.right = TreeNode(v); q.append(n.right!) }
        i += 1
    }
    return root
}

class Solution {
    func isSameTree(_ p: TreeNode?, _ q: TreeNode?) -> Bool {
        if p == nil && q == nil { return true }
        if p == nil || q == nil { return false }
        return p!.val == q!.val && isSameTree(p!.left, q!.left) && isSameTree(p!.right, q!.right)
    }
}

func checkBool(_ a: Bool, _ e: Bool, _ msg: String) {
    if a != e { fatalError("FAIL \(msg)") }
}

let sol = Solution()
checkBool(sol.isSameTree(fromLevel([1, 2, 3]), fromLevel([1, 2, 3])), true, "ex1")
checkBool(sol.isSameTree(fromLevel([1, 2]), fromLevel([1, nil, 2])), false, "ex2")
checkBool(sol.isSameTree(fromLevel([1, 2, 1]), fromLevel([1, 1, 2])), false, "ex3")
checkBool(sol.isSameTree(fromLevel([]), fromLevel([])), true, "empty")
checkBool(sol.isSameTree(fromLevel([1]), fromLevel([])), false, "one vs empty")
print("All tests passed!")

File: problems/07_trees/0100_same_tree/solution.dart
// Problem: 0100. Same Tree
// Difficulty: Easy

class TreeNode {
  int val; TreeNode? left, right;
  TreeNode([this.val = 0, this.left, this.right]);
}

TreeNode? fromLevel(List<int?> a) {
  if (a.isEmpty) return null;
  final root = TreeNode(a[0]!);
  final q = <TreeNode>[root]; var qi = 0, i = 1;
  while (qi < q.length && i < a.length) {
    final n = q[qi++];
    if (i < a.length && a[i] != null) { n.left = TreeNode(a[i]!); q.add(n.left!); }
    i++;
    if (i < a.length && a[i] != null) { n.right = TreeNode(a[i]!); q.add(n.right!); }
    i++;
  }
  return root;
}

bool isSameTree(TreeNode? p, TreeNode? q) {
  if (p == null && q == null) return true;
  if (p == null || q == null) return false;
  return p.val == q.val && isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
}

void _expect(bool a, bool e, String msg) {
  if (a != e) throw StateError('FAIL $msg');
}

void main() {
  _expect(isSameTree(fromLevel([1, 2, 3]), fromLevel([1, 2, 3])), true, 'ex1');
  _expect(isSameTree(fromLevel([1, 2]), fromLevel([1, null, 2])), false, 'ex2');
  _expect(isSameTree(fromLevel([1, 2, 1]), fromLevel([1, 1, 2])), false, 'ex3');
  _expect(isSameTree(fromLevel([]), fromLevel([])), true, 'empty');
  _expect(isSameTree(fromLevel([1]), fromLevel([])), false, 'one vs empty');
  print('All tests passed!');
}
